common/persistence/sql/sqlplugin/sqlite/plugin.go

```diff
@@ -22,6 +22,164 @@
+	"github.com/jmoiron/sqlx"
+
+	"github.com/uber/cadence/common/config"
+	"github.com/uber/cadence/common/persistence/sql/sqldriver"
+	"github.com/uber/cadence/common/persistence/sql/sqlplugin"
+)
+
 const (
 	PluginName = "sqlite"
 )
+
+// SQLite plugin provides an sql persistence storage implementation for sqlite database
+// Mostly the implementation reuses the mysql implementation
+// The plugin supports only in-memory sqlite database for now
+type plugin struct{}
+
+var _ sqlplugin.Plugin = (*plugin)(nil)
+
+// CreateDB wraps createDB to return an instance of sqlplugin.DB
+func (p *plugin) CreateDB(cfg *config.SQL) (sqlplugin.DB, error) {
+	return p.createDB(cfg)
+}
+
+// CreateAdminDB wraps createDB to return an instance of sqlplugin.AdminDB
+func (p *plugin) CreateAdminDB(cfg *config.SQL) (sqlplugin.AdminDB, error) {
+	return p.createDB(cfg)
+}
+
+// createDB create a new instance of DB
+func (p *plugin) createDB(cfg *config.SQL) (*DB, error) {
+	conns, err := sqldriver.CreateDBConnections(cfg, func(cfg *config.SQL) (*sqlx.DB, error) {
+		return p.createSingleDBConn(cfg)
+	})
+	if err != nil {
+		return nil, err
+	}
+	return NewDB(conns, nil, sqlplugin.DbShardUndefined, cfg.NumShards)
+}
+
+// createSingleDBConn creates a single database connection for sqlite
+func (p *plugin) createSingleDBConn(cfg *config.SQL) (*sqlx.DB, error) {
+	db, err := sqlx.Connect("sqlite3", buildDSN(cfg))
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.MaxConns > 0 {
+		db.SetMaxOpenConns(cfg.MaxConns)
+	}
+	if cfg.MaxIdleConns > 0 {
+		db.SetMaxIdleConns(cfg.MaxIdleConns)
+	}
+	if cfg.MaxConnLifetime > 0 {
+		db.SetConnMaxLifetime(cfg.MaxConnLifetime)
+	}
+
+	// Maps struct names in CamelCase to snake without need for DB struct tags.
+	db.MapperFunc(strcase.ToSnake)
+	return db, nil
+}
+
+// buildDSN builds the data source name for sqlite from config.SQL
+func buildDSN(cfg *config.SQL) string {
+
+	// by default, we use in-memory database if no database name is provided
+	var dsn = ":memory:"
+
+	// if database name is provided, then sqlite will use the file as the database
+	if cfg.DatabaseName != "" {
+		dsn = fmt.Sprintf("file:%s", cfg.DatabaseName)
+
+	}
+
+	if dsnAttrs := buildDSNAttrs(cfg); dsnAttrs != "" {
+		dsn += "?" + dsnAttrs
+	}
+
+	return dsn
+}
+
+const (
+	// if journal mode is not provided, we set it to WAL by default
+	// WAL mode allows readers and writers from different processes
+	// to access the database concurrently by default
+	// https://www.sqlite.org/wal.html
+	journalModeAttrName      = "_journal_mode"
+	journalModeAttrShortName = "_journal"
+	journalModeDefaultValue  = "WAL"
+)
+
+var (
+	journalModeAttrNames = []string{journalModeAttrName, journalModeAttrShortName}
+)
+
+// buildDSNAttrs builds the data source name attributes for sqlite from config.SQL
+// available attributes can be found here
+// https://github.com/mattn/go-sqlite3?tab=readme-ov-file#connection-string
+func buildDSNAttrs(cfg *config.SQL) string {
```

Comment:
Question: I guess url.Query doesn't work super well?

Reply:
Sorry for the intromission.

Assuming you're using my driver, I'm wondering what were the errors and what's the relationship with `mattn` documentation?

My driver does not support `mattn`'s query parameters (although it could).

Reply:
Hey Nuno, thank you for your comment, I will change these `mattn`'s query parameters to way how is described in your [lib](https://pkg.go.dev/github.com/ncruces/go-sqlite3/driver#hdr-Setting_PRAGMAs) 🙏